include/re_atomic.h

```diff
@@ -7,40 +7,717 @@
+#define RE_ATOMIC_WCHAR_T_LOCK_FREE __GCC_ATOMIC_WCHAR_T_LOCK_FREE
+#define RE_ATOMIC_SHORT_LOCK_FREE __GCC_ATOMIC_SHORT_LOCK_FREE
+#define RE_ATOMIC_INT_LOCK_FREE __GCC_ATOMIC_INT_LOCK_FREE
+#define RE_ATOMIC_LONG_LOCK_FREE __GCC_ATOMIC_LONG_LOCK_FREE
+#define RE_ATOMIC_LLONG_LOCK_FREE __GCC_ATOMIC_LLONG_LOCK_FREE
+#define RE_ATOMIC_POINTER_LOCK_FREE __GCC_ATOMIC_POINTER_LOCK_FREE
+
+#define re_memory_order_relaxed __ATOMIC_RELAXED
+#define re_memory_order_acquire __ATOMIC_ACQUIRE
+#define re_memory_order_release __ATOMIC_RELEASE
+#define re_memory_order_acq_rel __ATOMIC_ACQ_REL
+#define re_memory_order_seq_cst __ATOMIC_SEQ_CST
+
+#define re_atomic_store(_a, _v, _mo) \
+	__atomic_store_n(_a, _v, _mo)
+
+#define re_atomic_load(_a, _mo) \
+	__atomic_load_n(_a, _mo)
+
+#define re_atomic_exchange(_a, _v, _mo) \
+	__atomic_exchange_n(_a, _v, _mo)
+
+#define re_atomic_compare_exchange_strong(\
+	_a, _expected, _desired, _success_mo, _fail_mo) \
+	__atomic_compare_exchange_n(\
+		_a, _expected, _desired, 0, _success_mo, _fail_mo)
+
+#define re_atomic_compare_exchange_weak(\
+	_a, _expected, _desired, _success_mo, _fail_mo) \
+	__atomic_compare_exchange_n(\
+		_a, _expected, _desired, 1, _success_mo, _fail_mo)
+
+#define re_atomic_fetch_add(_a, _v, _mo) \
+	__atomic_fetch_add(_a, _v, _mo)
+
+#define re_atomic_fetch_sub(_a, _v, _mo) \
+	__atomic_fetch_sub(_a, _v, _mo)
+
+#define re_atomic_fetch_or(_a, _v, _mo) \
+	__atomic_fetch_or(_a, _v, _mo)
+
+#define re_atomic_fetch_xor(_a, _v, _mo) \
+	__atomic_fetch_xor(_a, _v, _mo)
+
+#define re_atomic_fetch_and(_a, _v, _mo) \
+	__atomic_fetch_and(_a, _v, _mo)
+
+/* MSVC Interlocked* intrinsics. This needs to go after clang to let clang-cl
+ * get handled above. */
 #elif defined(_MSC_VER)
-#define RE_ATOMIC volatile
-#define atomic_load(object) InterlockedOr((object), 0)
 
-/* C99 compiler builtin fallbacks */
-#elif defined(__clang__)
-#define __CLANG_ATOMICS
+#include <assert.h>
+#include <intrin.h>
+#include "re_types.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define RE_ATOMIC_BOOL_LOCK_FREE 2
+#define RE_ATOMIC_CHAR_LOCK_FREE 2
+#define RE_ATOMIC_WCHAR_T_LOCK_FREE 2
+#define RE_ATOMIC_SHORT_LOCK_FREE 2
+#define RE_ATOMIC_INT_LOCK_FREE 2
+#define RE_ATOMIC_LONG_LOCK_FREE 2
+#define RE_ATOMIC_LLONG_LOCK_FREE 2
+#define RE_ATOMIC_POINTER_LOCK_FREE 2
+
+/* These constants don't matter but for consistency they match
+ * values in std::memory_order from <atomic> in C++.
+ * There are specialized intrinsics for ARM and ARM64
+ * for different memory ordering types, but they are not used (yet) below. */
+#define re_memory_order_relaxed 0
+#define re_memory_order_acquire 2
+#define re_memory_order_release 3
+#define re_memory_order_acq_rel 4
+#define re_memory_order_seq_cst 5
 
-#elif defined(__GNUC__)
-#if (__GNUC__ << 16) + __GNUC_MINOR__ >= 0x40009
-#define __SYNC_ATOMICS
+#if defined(__INTEL_COMPILER)
+#define RE_COMPILER_BARRIER() __memory_barrier()
```

Comment:
Looks like this is deprecated?

https://docs.microsoft.com/en-us/cpp/intrinsics/readwritebarrier?view=msvc-170

Reply:
Well, there's no replacement in C, so I don't think they are going to remove it any time soon. And its use is fairly widespread, too.

Reply:
BTW, they are using it themselves:

https://github.com/microsoft/STL/blob/ef62d3fa0b8e4e2406b9bb74e916e1ca8a1df802/stl/inc/atomic#L32

Reply:
And it doesn't generate any warnings either:

https://gcc.godbolt.org/z/ejas9ff1f